src/daemon/daemon.cpp

```diff
@@ -690,14 +771,34 @@ mp::Daemon::Daemon(std::unique_ptr<const DaemonConfig> the_config)
             continue;
         }
 
-        auto mac_addr = spec.mac_addr;
-        if (mac_addr.empty())
+        auto interfaces = spec.interfaces;
+
+        // If there are no interfaces specified in spec, add to the instance the default one. If there are
+        // interfaces, check all of them for the mac address and generate it if missing.
+        if (interfaces.empty())
         {
-            mac_addr = mp::utils::generate_mac_address();
-            vm_instance_specs[name].mac_addr = mac_addr;
-            mac_addr_missing = true;
+            interfaces.push_back(mp::NetworkInterface{"default", generate_unused_mac_address()});
+        }
+        else
+        {
+            for (auto& iface : interfaces)
+            {
+                if (iface.mac_address.empty())
+                {
+                    iface.mac_address = generate_unused_mac_address();
+                    mac_addr_missing = true;
+                }
+                else
+                {
+                    // Make sure we have a correct mac address.
+                    // TODO: at some point, this check won't be needed. Remove, then.
+                    if (!mp::utils::valid_mac_address(iface.mac_address))
```

Comment:
Done. The validity of the MAC address is checked now when the user specifies a MAC or when the DB is constructed. I added here checking for repetition of addresses (can't do in other functions because it must be done inside the daemon). Do you think it's worth moving `load_db()` and `validate_extra_interfaces()` to the daemon in order to check for repetition there? This is possible, and will simplify the code.

Reply:
One thing that occurred to me is that `load_db` is used very early during object construction. Using methods in the constructor is dangerous, because it is easy to forget that class invariants may not hold just yet. If it's a very simple ctor calling an `init` function, I'd say that's ok. But using methods that call other methods in the earliest attribute initialization is something to avoid.

For instance, to make it work here you'd have to be sure that the collection of used macs was already initialized. IOW attribute order would have to be changed to avoid UB. But then it would be very easy for someone in the future to edit `load_db` or `validate_extra_interfaces` and use some other uninitialized attribute. So I'd say making those functions members would not be the right thing here, after all.

If you still wanted to have the macs be correct from initialization, you could instead extract methods into functions and make them receive the collection of used macs as a parameter. But I am not sure it is worth the trouble, all things considered.

Reply:
OK thanks. Yeah I agree with that approach. And nice thinking on the temporary mac set in case the instance is invalid. There are a few things that I think need a second look though. Will comment inline.